src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -485,42 +477,69 @@ internal override Task FlushAsync(CancellationToken cancellationToken = default)
             return Task.CompletedTask;
         }
 
-        public override ValueTask DisposeAsync()
-        {
-            // TODO: perform a graceful shutdown and wait for completion?
+        public override ValueTask DisposeAsync(CancellationToken cancellationToken) =>
+            DisposeAsync(cancellationToken, async: true, immediate: _state.SendState == SendState.Aborted);
 
-            Dispose(true);
-            return default;
-        }
+        public override void Dispose() =>
+            Dispose(immediate: _state.SendState == SendState.Aborted);
 
-        public override void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
+        ~MsQuicStream() =>
+            Dispose(immediate: true);
 
-        ~MsQuicStream()
+        private void Dispose(bool immediate)
         {
-            Dispose(false);
+            ValueTask t = DisposeAsync(cancellationToken: default, async: false, immediate);
+            Debug.Assert(t.IsCompleted);
+            t.GetAwaiter().GetResult();
         }
 
-        private void Dispose(bool disposing)
+        /// <param name="cancellationToken"></param>
+        /// <param name="async"></param>
+        /// <param name="immediate">When true, causes immediate disposal without waiting for peer ACKs.</param>
+        /// <returns></returns>
+        private async ValueTask DisposeAsync(CancellationToken cancellationToken, bool async, bool immediate)
         {
             if (_disposed)
             {
                 return;
             }
 
+            QUIC_STREAM_SHUTDOWN_FLAGS flags = immediate
+                ? (QUIC_STREAM_SHUTDOWN_FLAGS.GRACEFUL | QUIC_STREAM_SHUTDOWN_FLAGS.IMMEDIATE)
```

Comment:
This combination seems invalid... judging by the code, IMMEDIATE is only allowed in combination with both ABORT_SEND and ABORT_RECEIVE
https://github.com/microsoft/msquic/blob/c643c6275dfc506ec47c2067b676f26ba2368568/src/core/api.c#L909-L916